cloud/aws/aws_s3.cc

```diff
@@ -818,14 +819,80 @@ Status S3StorageProvider::CopyCloudObject(const std::string& bucket_name_src,
   return st;
 }
 
+namespace {
+class WritableFileStreamBuf : public std::streambuf {
+ public:
+  WritableFileStreamBuf(std::unique_ptr<WritableFileWriter> fileWriter)
+      : fileWriter_(std::move(fileWriter)) {}
+
+  ~WritableFileStreamBuf() {
+    fileWriter_->Flush();
+    fileWriter_->Close();
+  }
+
+ protected:
+  std::streamsize xsputn(const char* s, std::streamsize n) override {
+    auto st = fileWriter_->Append(rocksdb::Slice(s, n));
+    if (!st.ok()) {
+      return EOF;
+    }
+    return n;
+  }
+
+  int_type overflow(int_type ch) override {
+    if (traits_type::eq_int_type(ch, traits_type::eof())) {
+      return ch;
+    }
+    auto c = traits_type::to_char_type(ch);
+    auto r = xsputn(&c, 1);
+    if (r == EOF) {
+      return traits_type::eof();
+    }
+    // In case of success, the character put is returned
+    return ch;
+  }
+
+ private:
+  std::unique_ptr<WritableFileWriter> fileWriter_;
+};
+
+template <class T>
+class IOStreamWithUniquePtr : public std::iostream {
```

Comment:
This template could use some comments and probably a better name.  The "WithUniquePtr" does not really say what the type or purpose of the value is, nor does the name "s_".